dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+            sigmas. For example, in the isotropic case, the sigma of the
+            kernel will be $factor * (2 ^ i)$ where i = 0, 1, ..., n_scales
+            is the scale. The default is None.
+        options : dict, optional
+            extra optimization options. The default is None, implying
+            no extra options are passed to the optimizer.
+        """
+
+        self.metric = metric
+
+        if self.metric is None:
+            self.metric = MattesMIMetric()
+
+        if level_iters is None:
+            level_iters = [10000, 1000, 100]
+        self.level_iters = level_iters
+        self.levels = len(level_iters)
+        if self.levels == 0:
+            raise ValueError('The iterations list cannot be empty')
+
+        self.options = options
+        self.method = method
+        if ss_sigma_factor is not None:
+            self.use_isotropic = False
+            self.ss_sigma_factor = ss_sigma_factor
+        else:
+            self.use_isotropic = True
+            if factors is None:
+                factors = [4, 2, 1]
+            if sigmas is None:
+                sigmas = [3, 1, 0]
+            self.factors = factors
+            self.sigmas = sigmas
+
+    def _init_optimizer(self, static, moving, transform, x0,
+                        static_grid2space, moving_grid2space, prealign):
+        r"""Initializes the registration optimizer
+
+        Initializes the optimizer by computing the scale space of the input
+        images
+
+        Parameters
+        ----------
+        static: array, shape (S, R, C) or (R, C)
+            the image to be used as reference during optimization.
+        moving: array, shape (S, R, C) or (R, C)
+            the image to be used as "moving" during optimization. It is
+            necessary to pre-align the moving image to ensure its domain
+            lies inside the domain of the deformation fields. This is assumed
+            to be accomplished by "pre-aligning" the moving image towards the
+            static using an affine transformation given by the 'prealign'
+            matrix
+        transform: instance of Transform
+            the transformation with respect to whose parameters the gradient
+            must be computed
+        x0: array, shape (n,)
+            parameters from which to start the optimization. If None, the
+            optimization will start at the identity transform. n is the
+            number of parameters of the specified transformation.
+        static_grid2space: array, shape (dim+1, dim+1)
+            the voxel-to-space transformation associated with the static image
+        moving_grid2space: array, shape (dim+1, dim+1)
+            the voxel-to-space transformation associated with the moving image
+        prealign: string, or matrix, or None
+            If string:
+                'mass': align centers of gravity
+                'origins': align physical coordinates of voxel (0,0,0)
```

Comment:
> How will this differ from the transform given by the affines you have?

Oh! I see what the confusion may be: please remember that all operations are in physical space. The physical point corresponding to the (0,0,0) voxel in static and moving images may not be the same. If we specify `starting_affine='origins'` it will start with the translation needed to align both physical points. Of course, if both static and moving affines have the same translation, and we specify `starting_affine='origins'`, then the resulting pre-alignment matrix will be the identity.